flaky test: TestCheckpoint — chunker shrinks ChunkSize under CI load

Failing CI: https://github.com/block/spirit/actions/runs/25214878386 (MySQL 8.4 GA /w docker-compose)

```
--- FAIL: TestCheckpoint (4.50s)
    resume_test.go:227:
        Error: Not equal:
        expected: ChunkSize:1000, LowerBound:11001, UpperBound:12001
        actual:   ChunkSize:100,  LowerBound:4701,  UpperBound:4801
```

`TestCheckpoint` (`pkg/migration/resume_test.go:82`) sets `TargetChunkTime: 100ms` and asserts the watermark after exactly 11 chunks is `ChunkSize=1000, LowerBound=11001, UpperBound=12001`. Under CI load the optimistic chunker decided chunks were running slower than target and reduced `ChunkSize` from 1000 → 100, so 11 iterations advanced only ~3700 rows instead of ~10000 and the assertion fails.

The assertion bakes in a timing-dependent assumption about chunker auto-sizing. Likely fixes:

- Use a much higher `TargetChunkTime` (or disable dynamic sizing) for this test, or
- Relax the assertion to check monotonic advancement and ChunkSize ≥ some floor rather than an exact value.

Not a data race — pure timing flake. Failure on PR #754 is unrelated to that PR's metadata-lock change.

--- a/pkg/migration/resume_test.go
+++ b/pkg/migration/resume_test.go
@@ -103,12 +103,15 @@ func TestCheckpoint(t *testing.T) {
 
 	preSetup := func() *Runner {
 		r, err := NewRunner(&Migration{
-			Host:             cfg.Addr,
-			Username:         cfg.User,
-			Password:         &cfg.Passwd,
-			Database:         cfg.DBName,
-			Threads:          1,
-			TargetChunkTime:  100 * time.Millisecond,
+			Host:     cfg.Addr,
+			Username: cfg.User,
+			Password: &cfg.Passwd,
+			Database: cfg.DBName,
+			Threads:  1,
+			// Large TargetChunkTime so the optimistic chunker does not
+			// shrink ChunkSize under CI load — this test asserts an exact
+			// ChunkSize=1000 watermark, so dynamic resizing must not kick in.
+			TargetChunkTime:  time.Hour,
 			Table:            "cpt1",
 			Alter:            "ENGINE=InnoDB",
 			useTestThrottler: true,

--- a/pkg/migration/resume_test.go
+++ b/pkg/migration/resume_test.go
@@ -103,15 +103,12 @@ func TestCheckpoint(t *testing.T) {
 
 	preSetup := func() *Runner {
 		r, err := NewRunner(&Migration{
-			Host:     cfg.Addr,
-			Username: cfg.User,
-			Password: &cfg.Passwd,
-			Database: cfg.DBName,
-			Threads:  1,
-			// Large TargetChunkTime so the optimistic chunker does not
-			// shrink ChunkSize under CI load — this test asserts an exact
-			// ChunkSize=1000 watermark, so dynamic resizing must not kick in.
-			TargetChunkTime:  time.Hour,
+			Host:             cfg.Addr,
+			Username:         cfg.User,
+			Password:         &cfg.Passwd,
+			Database:         cfg.DBName,
+			Threads:          1,
+			TargetChunkTime:  100 * time.Millisecond,
 			Table:            "cpt1",
 			Alter:            "ENGINE=InnoDB",
 			useTestThrottler: true,
@@ -130,6 +127,17 @@ func TestCheckpoint(t *testing.T) {
 		require.NoError(t, r.changes[0].dropOldTable(t.Context()))
 		return r
 	}
+	// disableDynamicChunking turns off the optimistic chunker's adaptive
+	// resizing so this test sees a stable ChunkSize=1000 regardless of
+	// per-chunk timing under CI load. The test asserts watermark contents,
+	// not chunker auto-sizing behavior.
+	disableDynamicChunking := func(c table.Chunker) {
+		t.Helper()
+		setter, ok := c.(interface{ SetDynamicChunking(bool) })
+		require.True(t, ok, "copyChunker does not expose SetDynamicChunking")
+		setter.SetDynamicChunking(false)
+	}
+
 	r := preSetup()
 	// migrationRunner.Run usually calls r.Setup() here.
 	// Which first checks if the table can be restored from checkpoint.
@@ -137,6 +145,7 @@ func TestCheckpoint(t *testing.T) {
 	require.Error(t, r.resumeFromCheckpoint(t.Context()))
 	// So we proceed with the initial steps.
 	require.NoError(t, r.newMigration(t.Context()))
+	disableDynamicChunking(r.copyChunker)
 
 	// Now we are ready to start copying rows.
 	// Instead of calling r.copyRows() we will step through it manually.
@@ -196,6 +205,7 @@ func TestCheckpoint(t *testing.T) {
 	// Start the binary log feed just before copy rows starts.
 	// replClient.Run() is already called in resumeFromCheckpoint.
 	require.NoError(t, r.resumeFromCheckpoint(t.Context()))
+	disableDynamicChunking(r.copyChunker)
 	// This opens the table at the checkpoint (table.OpenAtWatermark())
 	// which sets the chunkPtr at the LowerBound. It also has to position
 	// the watermark to this point so new watermarks "align" correctly.

--- a/pkg/table/chunker_optimistic.go
+++ b/pkg/table/chunker_optimistic.go
@@ -207,7 +207,10 @@ func (t *chunkerOptimistic) Open() (err error) {
 	return t.open()
 }
 
-func (t *chunkerOptimistic) setDynamicChunking(newValue bool) {
+// SetDynamicChunking enables (true) or disables (false) the optimistic
+// chunker's dynamic chunk-size adaptation. Disabling is intended for tests
+// that need deterministic chunk sizes; production callers should leave it on.
+func (t *chunkerOptimistic) SetDynamicChunking(newValue bool) {
 	t.Lock()
 	defer t.Unlock()
 	t.disableDynamicChunker = !newValue

--- a/pkg/table/chunker_optimistic_test.go
+++ b/pkg/table/chunker_optimistic_test.go
@@ -30,7 +30,7 @@ func TestOptimisticChunkerBasic(t *testing.T) {
 		ChunkerTarget: ChunkerDefaultTarget,
 		logger:        slog.Default(),
 	}
-	chunker.setDynamicChunking(false)
+	chunker.SetDynamicChunking(false)
 
 	assert.NoError(t, t1.PrimaryKeyIsMemoryComparable())
 	t1.keyColumnsMySQLTp[0] = "varchar"
@@ -90,7 +90,7 @@ func TestLowWatermark(t *testing.T) {
 		lowerBoundWatermarkMap: make(map[string]*Chunk, 0),
 		logger:                 slog.Default(),
 	}
-	chunker.setDynamicChunking(false)
+	chunker.SetDynamicChunking(false)
 
 	assert.NoError(t, chunker.Open())
 
@@ -318,7 +318,7 @@ func TestOptimisticPrefetchChunking(t *testing.T) {
 		ChunkerTarget: time.Second,
 		logger:        slog.Default(),
 	}
-	chunker.setDynamicChunking(true)
+	chunker.SetDynamicChunking(true)
 	assert.NoError(t, chunker.Open())
 	assert.False(t, chunker.chunkPrefetchingEnabled)
 
@@ -354,7 +354,7 @@ func TestOptimisticChunkerReset(t *testing.T) {
 		lowerBoundWatermarkMap: make(map[string]*Chunk, 0),
 		logger:                 slog.Default(),
 	}
-	chunker.setDynamicChunking(false)
+	chunker.SetDynamicChunking(false)
 
 	// Test that Reset() fails when chunker is not open
 	err := chunker.Reset()

